spatialdata/_core/_spatialdata.py

```diff
@@ -48,163 +50,301 @@ class SpatialData:
+        ndim = len(get_dims(labels))
+        if ndim == 2:
+            Label2D_s.validate(labels)
+            self._labels[name] = labels
+        elif ndim == 3:
+            Label3D_s.validate(labels)
+            self._labels[name] = labels
+        else:
+            raise ValueError(f"Only yx and zyx labels supported, got {ndim} dimensions")
+
+    def _add_polygons_in_memory(self, name: str, polygons: GeoDataFrame) -> None:
+        Polygon_s.validate(polygons)
+        self._polygons[name] = polygons
+
+    def _add_shapes_in_memory(self, name: str, shapes: AnnData) -> None:
+        Shape_s.validate(shapes)
+        self._shapes[name] = shapes
+
+    def _add_points_in_memory(self, name: str, points: pa.Table) -> None:
+        Point_s.validate(points)
+        self._points[name] = points
+
+    def is_backed(self) -> bool:
+        """Check if the data is backed by a Zarr storage or it is in-memory."""
+        return self.path is not None
+
+    def _init_add_element(self, name: str, element_type: str, overwrite: bool) -> zarr.Group:
+        if self.path is None:
+            # in the future we can relax this, but this ensures that we don't have objects that are partially backed
+            # and partially in memory
+            raise RuntimeError(
+                "The data is not backed by a Zarr storage. In order to add new elements after "
+                "initializing a SpatialData object you need to call SpatialData.write() first"
+            )
+        store = parse_url(self.path, mode="r+").store
+        root = zarr.group(store=store)
+        assert element_type in ["images", "labels", "points", "polygons", "shapes"]
+        # not need to create the group for labels as it is already handled by ome-zarr-py
+        if element_type != "labels":
+            if element_type not in root:
+                elem_group = root.create_group(name=element_type)
+            else:
+                elem_group = root[element_type]
+        if overwrite:
+            if element_type == "labels":
+                if element_type in root:
+                    elem_group = root[element_type]
+            if name in elem_group:
+                del elem_group[name]
+        else:
+            # bypass is to ensure that elem_group is defined. I don't want to define it as None but either having it
+            # or not having it, so if the code tries to access it and it should not be there, it will raise an error
+            bypass = False
+            if element_type == "labels":
+                if element_type in root:
+                    elem_group = root[element_type]
+                else:
+                    bypass = True
+            if not bypass:
+                if name in elem_group:
+                    raise ValueError(f"Element {name} already exists, use overwrite=True to overwrite it")
+
+        if element_type != "labels":
+            return elem_group
+        else:
+            return root
+
+    def add_image(
```

Comment:
please add a doctsring. at a minimum, please add a description of what the function does.

Reply:
done